sourcecred/sourcecred
# Discord: fetch backwards and add a beginningDate config

# Description
Discord's rate limiting is getting harsher, and loading the full history is becoming too long even for smaller communities. This PR attempts to address this by allowing users to set a beginningDate, and then fetching backwards until that date is reached and then stopping. For example, users could choose to only load messages after 1/1/2022.


# Test Plan
1. Add `console.log(new Date(messages[0]?.timestamp))` on line 180 of the fetcher.
2. Optionally comment out lines 117-126 to disable reaction fetching for quicker testing.
3. yarn build
4. checkout sourcecred/cred/gh-pages
5. scdev load and verify that only recent messages are pulled (the cache is respected and not re-pulled)
6. yarn clean
7. scdev load and verify messages are pulled in reverse order
8. ctrl + C to interrupt after a channel has loaded 3 pages
9. scdev load and verify that a recent page or two is reloaded, but then it skips most of what is in the cache to resume loading before what is in the cache.
10. yarn clean
11. add `"beginningDate": "1/22/2022"` in discord/config.js
12. scdev load and verify that messages are not loaded once the beginning date is reached
